pmap.c

```diff
@@ -0,0 +1,107 @@
+#include <pmap.h>
+#include <malloc.h>
+#include <libkern.h>
+#include <mips/cpu.h>
+#include <tlb.h>
+
+#define PTE_INDEX(x)  (((x) & 0xfffff000) >> 12)
+#define PDE_INDEX(x) (((x) & 0xffc00000) >> 22)
+
+pmap_t *global_pmap;
+
+void pmap_init(pmap_t *pmap)
+{
+    pmap->pte = (pte_t*)PTE_BASE;
+    pmap->pde_page = vm_phys_alloc(0);
+    pmap->pde = (pte_t*)pmap->pde_page->virt_addr;
+    TAILQ_INIT(&pmap->pte_pages);
+}
+
+void pmap_delete(pmap_t *pmap)
+{
+    while(!TAILQ_EMPTY(&pmap->pte_pages))
+    {
+        vm_phys_free(TAILQ_FIRST(&pmap->pte_pages));
+        TAILQ_REMOVE(&pmap->pte_pages, TAILQ_FIRST(&pmap->pte_pages),vm.pt.list);
+    }
+    vm_phys_free(pmap->pde_page);
+}
+
+static void pde_map(pmap_t *pmap, vaddr_t vaddr)
+{
+    uint32_t pde_index = PDE_INDEX(vaddr);
+    if( (pmap->pde[pde_index] & V_MASK) == 0) /* part of page table isn't located in memory */
+    {
+        vm_page_t *pg = vm_phys_alloc(0);
+        TAILQ_INSERT_TAIL(&pmap->pte_pages, pg, vm.pt.list);
+
+        SET_PADDR(pmap->pde[pde_index],pg->phys_addr);
+        SET_V(pmap->pde[pde_index],1);
+        SET_D(pmap->pde[pde_index],1);
+    }
+
+    /* make sure proper address is in tlb */
+    pde_index &= ~1;
+    tlbhi_t entryhi = (PTE_BASE + pde_index*PAGESIZE) | pmap->asid;
+    tlblo_t entrylo0 = pmap->pde[pde_index];
+    tlblo_t entrylo1 = pmap->pde[pde_index+1];
+    uint32_t pmask = 0;
+    tlb_rwr2(entryhi, entrylo0, entrylo1, pmask);
+}
+
+static void pt_map(pmap_t *pmap, vm_addr_t vaddr, vm_paddr_t paddr, uint32_t flags)
+{
+    pde_map(pmap, vaddr);
+    pte_t entry = flags;
+    uint32_t pt_index = PTE_INDEX(vaddr);
+    SET_PADDR(entry ,paddr);
+    pmap->pte[pt_index] = entry;
+
+    /* invalidate proper entry in tlb */
+    tlbhi_t entryhi = (vaddr & 0xfffffc00) | pmap->asid;
+    tlb_inval(entryhi);
+}
+
+void pmap_map_range(pmap_t *pmap, vm_addr_t vaddr, vm_paddr_t paddr, size_t npages, uint32_t flags)
+{
+    for(size_t i = 0; i < npages; i++)
+        pt_map(pmap, vaddr + i*PAGESIZE, paddr + i*PAGESIZE, flags);
+}
+
+void pmap_set_global_pmap(pmap_t *pmap)
```

Comment:
If you wrote a setter function, please write a getter as well - apparently accessing global variable directly is not the way in this case. Name those functions `set_active_pmap` and `get_active_pmap`.